mopidy/internal/path.py

```diff
@@ -205,11 +213,16 @@ def find_mtimes(root, follow=False):
 
 def is_path_inside_base_dir(path, base_path):
     if not isinstance(path, bytes):
-        raise ValueError('path is not a bytestring')
+        raise TypeError('path is not a bytestring')
     if not isinstance(base_path, bytes):
-        raise ValueError('base_path is not a bytestring')
+        raise TypeError('base_path is not a bytestring')
+
+    if compat.PY2:
+        path_separator = os.sep
+    else:
+        path_separator = os.sep.encode()
 
-    if path.endswith(os.sep):
+    if path.endswith(path_separator):
         raise ValueError('Path %s cannot end with a path separator'
```

Comment:
Won't this mix of unicode (string literal) and bytes (`path`) potentially be a problem?

Reply:
Yeah, it will cause non-ascii characters in the path to be printed as escaped values.

```
>>> 'Path %s cannot' % '/testå'.encode('utf-8')
"Path b'/test\\xc3\\xa5' cannot"
```

Shouldn't we use `str` instead of `bytes` for python 3?

Reply:
Paths in Mopidy today are bytes, and we need to keep them that way. In the short run to make a gradual porting possible, where the Python 2 test suite passes all the way. In the long run, because most file systems can have any byte sequence as file names, and experience shows that many music collections has a strange mix of file name encodings.

I think we can improve on the current situation by becoming a lot more explicit if something is bytes or text, which Python 3 in general will help with. Optional typing, once we're on Python 3, can also help. And I'd also like to introduce an `URI` class that we can use to pass paths around inside Mopidy as URI-encoded `file://` paths.

Reply:
> I'll change it from %s to %r.

That doesn't help, as far as I can see? %s and %r prints the exact same thing for the bytes I've tested.

> In the long run, because most file systems can have any byte sequence as file names, and experience shows that many music collections has a strange mix of file name encodings.

Python has an escape sequence which is used for such paths. E.g. `os.listdir` will return a `str` if you pass a `str` into it (or if you don't pass anything, it will only return `bytes` if you pass in `bytes`), and `os.fsencode` and `os.fsdecode` can be used to convert between escaped `str` and `bytes`.

Though, even if you use these escaped `str`s, you will have to convert them before you can print the path.

> I think we can improve on the current situation by becoming a lot more explicit if something is bytes or text, which Python 3 in general will help with. Optional typing, once we're on Python 3, can also help.

Agreed.

Reply:
No, switching to `%r` doesn't help, it just makes the `b'...'` rendering explicitly requested. Good enough for now, was my thinking.

---

I built my own `Path` class a few days ago, which internally stores the path as a unicode object with the input URI-encoded. It has `path.as_bytes() -> bytes` which returns the exact same bytes as it was instantiated with, and `path.for_display() -> str`, using "replace" as the UTF-8 decoding error strategy.

During the work on this, I found `os.fsencode()` and `os.fsdecode()`, which you mention, but they were added in Python 3.2, so they're not very helpful at this stage. This `Path` class I created basically reimplements them internally with the help of `sys.getfilesystemencoding()`.

---

Looking at this a couple of days later, I feel like I'm inventing a lousy version of something that already exists...

**Python 3 has the `pathlib` module, and PyPI has the `pathlib2` backport.**

Testing them out, it seems they behave quite well on files containing invalid UTF-8 byte sequences in the file name.

On Python 2:

``` python
>>> from pathlib2 import Path
>>> p = Path('.')
>>> list(p.glob('foo*'))
[PosixPath('foo\xc3('), PosixPath('foo\xc3(\xc3')]
>>> f = list(p.glob('foo*'))[1]
>>> bytes(f)
'foo\xc3(\xc3'   # The correct bytes
>>> str(f)
'foo\xc3(\xc3'   # Exactly the same, as bytes == str on Python 2
>>> bytes(f).decode('utf-8', 'replace')
u'foo\ufffd(\ufffd'  # Unicode, ready for display, with invalid bytes replaced
>>> print(bytes(f).decode('utf-8', 'replace'))
foo�(�
```

On Python 3:

``` python
>>> from pathlib import Path
>>> p = Path('.')
>>> list(p.glob('foo*'))
[PosixPath('foo\udcc3('), PosixPath('foo\udcc3(\udcc3')]    # Surrogate escapes in action
>>> f = list(p.glob('foo*'))[1]
>>> bytes(f)
b'foo\xc3(\xc3'    # The correct bytes, just with Python 3's b'' prefix
>>> str(f)
'foo\udcc3(\udcc3'    # Surrogate escapes again. This isn't "printable".
>>> bytes(f).decode('utf-8', 'replace')
'foo�(�'    # Unicode, ready for display, with invalid bytes replaced
>>> print(bytes(f).decode('utf-8', 'replace'))
foo�(�
```

---

My plan now is to:

1. Merge this PR as is.
2. Do a spike on replacing Mopidy's custom path stuff with pathlib (and `pathlib2` for Python 2 during the migration).

I think that if we end up with:

- passing `pathlib.Path` objects around when we deal with paths,
- when we display `Path` objects, either use `repr()` or make sure to do the replace-dance,
- URI-encode properly when we convert to `file://` or `local:` URIs,

then we should be in a very good place with regards to path handling, both for Python 3 and for Windows.